File: ios/Classes/barcode/SupyBarcodeScannerViewFactory.swift
import Flutter
import UIKit

/// Registers the embedded barcode scanner PlatformView.
///
/// View-type identifier and per-view channel scheme are defined in
/// `docs/ARCHITECTURE.md`. The Android counterpart is
/// `SupyBarcodeScannerViewFactory.kt`.
final class SupyBarcodeScannerViewFactory: NSObject, FlutterPlatformViewFactory {

  static let viewTypeId = "io.supy.scanner/v1/barcode_view"

  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let params = args as? [String: Any?]
    return SupyBarcodeScannerView(
      frame: frame,
      viewId: viewId,
      creationParams: params,
      messenger: messenger
    )
  }

  // Matches the Android `StandardMessageCodec` codec used by the Dart side.
  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}

File: android/src/main/kotlin/io/supy/scanner/barcode/SupyBarcodeScannerViewFactory.kt
package io.supy.scanner.barcode

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * Factory for [SupyBarcodeScannerView] PlatformViews.
 *
 * Registered against the view-type identifier `io.supy.scanner/v1/barcode_view`
 * — must match the Dart side in `SupyBarcodeScannerView` (Dart).
 */
class SupyBarcodeScannerViewFactory(
    private val messenger: BinaryMessenger,
    private val activityHolder: ActivityHolder,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val creationParams = args as? Map<String, Any?>
        return SupyBarcodeScannerView(
            context = context,
            viewId = viewId,
            creationParams = creationParams,
            messenger = messenger,
            activityHolder = activityHolder,
        )
    }

    companion object {
        const val VIEW_TYPE_ID = "io.supy.scanner/v1/barcode_view"
    }
}
